tests/protocol/test_metadata_server_migration_util.py

```diff
@@ -0,0 +1,131 @@
+    @patch('azurelinuxagent.common.conf.get_lib_dir')
+    def test_cleanup_metadata_server_artifacts_does_not_throw_with_no_metadata_certs(self, mock_get_lib_dir, mock_enable_firewall):
+        mock_get_lib_dir.return_value = tempfile.gettempdir()
+        mock_enable_firewall.return_value = False
+        osutil = MagicMock()
+        migration_util.cleanup_metadata_server_artifacts(osutil)
+
+    @patch('azurelinuxagent.common.conf.enable_firewall')
+    @patch('azurelinuxagent.common.conf.get_lib_dir')
+    @patch('os.getuid')
+    def test_cleanup_metadata_server_artifacts_firewall_enabled(self, mock_os_getuid, mock_get_lib_dir, mock_enable_firewall):
+        # Setup Certificate Files
+        dir = tempfile.gettempdir()
+        metadata_server_transport_prv_file = os.path.join(dir, _LEGACY_METADATA_SERVER_TRANSPORT_PRV_FILE_NAME)
+        metadata_server_transport_cert_file = os.path.join(dir, _LEGACY_METADATA_SERVER_TRANSPORT_CERT_FILE_NAME)
+        metadata_server_p7b_file = os.path.join(dir, _LEGACY_METADATA_SERVER_P7B_FILE_NAME)
+        open(metadata_server_transport_prv_file, 'w').close()
+        open(metadata_server_transport_cert_file, 'w').close()
+        open(metadata_server_p7b_file, 'w').close()
+
+        # Setup Mocks
+        mock_get_lib_dir.return_value = dir
+        mock_enable_firewall.return_value = True
+        fixed_uid = 0
+        mock_os_getuid.return_value = fixed_uid
+        osutil = MagicMock()
+
+        # Run
+        migration_util.cleanup_metadata_server_artifacts(osutil)
+
+        # Assert files deleted
+        self.assertFalse(os.path.exists(metadata_server_transport_prv_file))
+        self.assertFalse(os.path.exists(metadata_server_transport_cert_file))
+        self.assertFalse(os.path.exists(metadata_server_p7b_file))
+
+        # Assert Firewall rule calls
+        osutil.remove_rules_files.assert_called_once()
+        osutil.remove_firewall.assert_called_once_with(dst_ip=_KNOWN_METADATASERVER_IP, uid=fixed_uid)
+        osutil.enable_firewall.assert_called_once_with(dst_ip=KNOWN_WIRESERVER_IP, uid=fixed_uid)
+
+    @patch('azurelinuxagent.common.conf.enable_firewall')
+    @patch('azurelinuxagent.common.conf.get_lib_dir')
+    @patch('os.getuid')
+    def test_cleanup_metadata_server_artifacts_firewall_disabled(self, mock_os_getuid, mock_get_lib_dir, mock_enable_firewall):
+        # Setup Certificate Files
+        dir = tempfile.gettempdir()
+        metadata_server_transport_prv_file = os.path.join(dir, _LEGACY_METADATA_SERVER_TRANSPORT_PRV_FILE_NAME)
+        metadata_server_transport_cert_file = os.path.join(dir, _LEGACY_METADATA_SERVER_TRANSPORT_CERT_FILE_NAME)
+        metadata_server_p7b_file = os.path.join(dir, _LEGACY_METADATA_SERVER_P7B_FILE_NAME)
+        open(metadata_server_transport_prv_file, 'w').close()
+        open(metadata_server_transport_cert_file, 'w').close()
+        open(metadata_server_p7b_file, 'w').close()
+
```

Comment:
You could refactor all the common code into a `_setup_metadata_protocol_setup`-type method.

Reply:
Or just override the `setUp` method of `AgentTestCase`.

Reply:
@byue - Dont worry about this comment - You can do it in a different PR.